### ISSUE TITLE
Add Shopify Button Functionality to Gallery Pages

### ISSUE THREAD

**Issue by RassanGrant** [Norstone/norstone.github.io#216]
Hi Vicky - 

The other thing we'll need to get done before going live with the Shopify project is the dynamic Shopify ID button insertion on the application galleries, like this one: https://www.norstoneusa.com/gallery/application/fireplace/

Below is a screenshot of what I am trying to accomplish. Similar to how the product image/link change when the image changes, we also need the shopify buy sample link id to change dynamically. I think this should actually be fairly straight forward because all the images are already tagged, we would just need to include the ID with the tags.

- You will need to decrease the size of the social icons and decrease the spacing between the additional galleries drop-downs in order to make room for the Buy A Sample button. 

![app-shopping-cart-page](https://user-images.githubusercontent.com/18408413/33788678-8eecbbee-dc41-11e7-8a10-c2d647598106.jpg)

**Comment by vleong2332** [comment]
@RassanGrant, I'll have to look into it a bit to come up with an estimate, but it's certainly doable.

Since we're adding an element (button), how do you want the elements to be aligned?

**Comment by RassanGrant** [comment]
Hey Vicky - I think the elements should look as close to what I have created in the attachment as possible. To do this, you'll definitely need to shrink the share icons and you shorten the space between the two filter drop downs.

**Comment by vleong2332** [comment]
I see. Do you want the right and left columns to be the same length? In the picture the right column is shifted lower slightly.

**Comment by RassanGrant** [comment]
Ah, that was unintentional - It should be on par

**Comment by vleong2332** [comment]
@RassanGrant I estimate 40 minutes to 1 hour for this item. I can do it within 24 hours too.

**Comment by RassanGrant** [comment]
Hey Vicky - I see you've put a little time into this one. Not sure if you've identified it, but one of the issues I was having while checking out your changes were that the image gallery failed to load. See screen grab. It did work when I hit shift + referesh so I'm not sure what that's doing differently. Also I noticed that the add-to-cart only worked for the Ochre Rock Panel, but I assume that's just because you're still testing. Just wanted to post a round of feedback here - Thanks, @vleong2332!

<img width="1004" alt="2017-12-13 08 57 12 am" src="https://user-images.githubusercontent.com/18408413/33942464-fd409e58-dfe3-11e7-8fea-683c79f828d1.png">

**Comment by vleong2332** [comment]
@RassanGrant Correct. This turned out to be a little bit tricker than expected. I started with the Ochre rock panel first to see how things go. I already burned up a whole hour on this so I want to ask for more time. If you like the progress, I will make it work for other products and then adjust the layout on that column. I estimate another 30 minutes to 1 hour for that.

**Comment by vleong2332** [comment]
@RassanGrant I spent another 45 minutes to fix the layout and the different product ID. Let me know what you think.

I think the product ID for the charcoal rock is wrong. Could you double check that?

**Comment by RassanGrant** [comment]
Hey Vicky - Good eye on the Charcoal - The product ID is id: 564268498987.

Two things: 1. My Chrome is still having trouble rendering the image gallery unless I do a hard refresh on the page - Are you seeing that at all on your side?
2. I notice that when the button changes, it reloads the button each time, which leads to a brief delay while it loads. Ideally I'd like to see the button stay static on the page (like the product card) where the only thing that changes is the product ID behind the scenes. Can you accomplish that?

**Comment by vleong2332** [comment]
@RassanGrant, I'm not having trouble with the image gallery. The next time it happens, could you look at the chrome console and see if any error popped up? I'll keep an eye out for that behavior on my end.

Right, so for both the product image and the button, they get updated (re-initialize) when the currently selected gallery image changes. If I recall correctly, the product image actually gets reinitialized (requesting new image) every time you select a new gallery image. But since the load time for the image is relatively fast on a decent connection, we barely notice the change.

For the button, it only reinitialize if the product type in the currently selected gallery image is different from the last one. The reason you see it reloads because I have to destroy the element and then reinitialize it with a new product id. Since we're talking to a 3rd party API, the delay is noticeable.

Let me think of a strategy to minimize or even eliminate that delay. I'll update you on my thoughts throughout the day.

**Comment by RassanGrant** [comment]
Sounds great @vleong2332 - Thanks!

**Comment by vleong2332** [comment]
Hi, Rassan. I think we need to be a little bit creative to solve this problem. One way I thought about was to create a totally different button for a different product type. When a gallery image with a different product type is selected, we can look for an existing button with that product id. If it's there, we simply swap the buttons without the user noticing. If it's not there, then we'll have to initialize the button and swap it when it's done initializing. I'm not sure how seamless I can make the second case (initializing the button for the first time), but I can try.

As for the time, since this is going to be pretty much JavaScript work-around, I'm estimating 1-2 hours. Let me know what you think.

**Comment by RassanGrant** [comment]
Hey @vleong2332 - Let's ping @albertvolkman for his feedback on this. The goal should be to create a seamless user experience so that when they are flipping through the images the button appears to remain on the page. I think what you're proposing makes sense, but I'd like to get Albert's opinion on it.

**Comment by vleong2332** [comment]
Sure. I would love his input as well.

**Comment by albertvolkman** [comment]
Am I correct assuming that this is the current code?

https://github.com/Norstone/norstone.github.io/compare/shopify?expand=1

**Comment by vleong2332** [comment]
Yes, that's the branch.

**Comment by albertvolkman** [comment]
Why are we recreating the button each time instead of simply switching out the link? I'm not familiar with the Shopify API... is this required?

**Comment by vleong2332** [comment]
That was our first thought, but I couldn't find a way to "just switch the id" of an initialized button. https://github.com/Norstone/norstone.github.io/blob/shopify/_js/shopify_cart.js#L258 creates the button. In the created markup (iframe), there is no reference to the ID number anymore.

**Comment by vleong2332** [comment]
I need to double check, but I don't think the button is a link. It's more of a script triggered by a click.
